File: leetcode008.py
class Solution:
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        ls = len(str)
        if ls == 0:
            return 0
        i = 0
        res = 0
        t = 1
        while i < ls and str[i] == ' ':
            i = i + 1
        if i<ls and (str[i] == '-' or str[i] == '+'):
            if str[i] == '-':
                t = -1
            elif str[i] == '+':
                t = 1
            i = i + 1
        while i < ls and ord(str[i])>=ord('0') and ord(str[i])<=ord('9'):
            res = res * 10 + (ord(str[i]) - ord('0'))
            i = i + 1
        res = res * t
        if res > 2**31-1:
            return 2**31-1
        if res < -2**31:
            return -2**31
        return res

if __name__=="__main__":
    s = Solution()
    print(s.myAtoi(' '))

File: leetcodec++/leetcode008.cpp
//
// Created by 黄敏辉 on 2019-06-13.
//

class Solution {
public:
    int myAtoi(string str) {
        long res = 0;

        if (str.empty())
            return res;

        int sign = 1;
        int ind = str.find_first_not_of(' ');

        if (ind == -1)
            return res;

        if (str[ind] == '+' || str[ind] == '-') sign = str[ind++] == '+' ? 1 : -1;

        while (ind < str.length() && isdigit(str[ind])) {
            res = res * 10 + (str[ind++] - '0');
            if (res * sign > INT_MAX) return INT_MAX;
            if (res * sign < INT_MIN) return INT_MIN;
        }

        return res * sign;

    }
};